src/go/k8s/internal/client/kubernetes/client.go

```diff
@@ -0,0 +1,77 @@
+// Package kubernetes holds logic for doing server side applies with a controller-runtime client.
+package kubernetes
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// Client adds some helper methods to a standard controller runtime client.
+type Client interface {
+	client.Client
+	ClearFinalizer(ctx context.Context, o client.Object, finalizer string) error
+	Apply(ctx context.Context, o any, owner client.FieldOwner) error
+	ApplyStatus(ctx context.Context, o any, owner client.FieldOwner) error
+}
+
+// KubernetesClient wraps a controller-runtime client, extending it with
+// an Apply method that does server side apply with a typed ApplyConfiguration builder.
+type KubernetesClient struct {
+	client.Client
+}
+
+// Wrap wraps a standard controller runtime client with some helpers for server-side apply
+// of *ApplyConfiguration objects.
+func Wrap(c client.Client) *KubernetesClient {
+	return &KubernetesClient{
+		Client: c,
+	}
+}
+
+// ClearFinalizer clears a finalizer from the given object. Because finalizers are generally cleared
+// only when a resource is no longer managed, this should be the final step in any reconciliation loop.
+// Additionally, due to finalizers being annotated with patchStrategy=merge, we can't remove a finalizer
+// via a typical server-side apply, so instead we send a full Update to the api server.
+func (c *KubernetesClient) ClearFinalizer(ctx context.Context, o client.Object, finalizer string) error {
+	if controllerutil.RemoveFinalizer(o, finalizer) {
+		return c.Update(ctx, o)
+	}
+
+	return nil
+}
+
+// Apply issues a server-side apply request using conversion routines to an unstructured object.
+// Due to the limitations of generated ApplyConfiguration objects not having a consistent interface
+// we pass an arbitrary interface as an argument here. That said, callers should only use this function
+// with ApplyConfiguration objects.
+func (c *KubernetesClient) Apply(ctx context.Context, o any, owner client.FieldOwner) error {
```

Comment:
One of the quirks of the client.Client that I've grown accustomed to is that it modifies the provided object in place and allows callers to read back the modified object. I don't love that that convention is being broken here.

I wonder if we should take a whack at something like https://github.com/kubernetes/client-go/blob/master/gentype/type.go#L310. I suspect we won't find a lot of the pain points until we're in a post flux world.

There's also another client wrapper in the helm chart that we'll want to _reconcile_ with soon.

Reply:
So, I could remove the client wrapper if we want to introduce our own custom `Patch` implementation to specify `Data` based on an `ApplyConfiguration` [similar to this](https://github.com/Jefftree/controller-runtime/blob/ab9ceff6322d8838b2bed5e5cf250d37730bdca1/pkg/client/patch.go#L236-L247) which wasn't actually merged into the controller-runtime. I believe that would also have the benefit of unmarshaling the applied resource back onto the `client.Object` we pass it. Thoughts?

Reply:
I just went ahead and implemented roughly that -- an `ApplyPatch` helper that returns a `client.Patch` implementation that just delegates to whatever you pass it to provide the underlying data. [Here's where it gets used](https://github.com/Jefftree/controller-runtime/blob/9e78e653228851684b6b9cdabe5aae8559fe3722/pkg/client/typed_client.go#L112-L132) in controller runtime, so it's pretty much exactly going to be what you linked above.